### PR TITLE
Boskos ranch: Use crd types

### PR DESCRIPTION
**Pull request by alvaroaleman**
This PR changes boskos ranch to use the crd types, working towards https://github.com/kubernetes/test-infra/issues/16207 and working towards https://github.com/kubernetes/test-infra/issues/15557 by not short circuiting kubes optimistic concurrency mechanism anymore.

Boskos ranch and storage are used a lot in other packages tests, which is why this PR grew so much in size.

### CODE REVIEW

File: boskos/handlers/server_client_test.go [hunk 1]
```diff
@@ -283,14 +286,15 @@ func TestClientServerUpdate(t *testing.T) {
 				t.Fatalf("tc: %s - errors don't match, expected %v, received\n %v", tc.name, tc.err, err)
 			}
 			receivedRes, _ := r.Storage.GetResource(tc.resource.Name)
-			if !reflect.DeepEqual(receivedRes.UserData.ToMap(), tc.expected.UserData.ToMap()) {
-				t.Errorf("tc: %s - resources user data should match. Expected \n%v, received \n%v", tc.name, tc.expected.UserData.ToMap(), receivedRes.UserData.ToMap())
+			if !reflect.DeepEqual(receivedRes.Status.UserData.ToMap(), tc.expected.Status.UserData.ToMap()) {
+				t.Errorf("tc: %s - resources user data should match. Expected \n%v, received \n%v", tc.name, tc.expected.Status.UserData.ToMap(), receivedRes.Status.UserData.ToMap())
 			}
 			// Hack: remove UserData to be able to compare since we already compared it before.
```
Comment (ixdy) [inline boskos/handlers/server_client_test.go]: I wonder if this is still necessary.

File: boskos/ranch/ranch_test.go [hunk 2]
```diff
@@ -455,6 +455,9 @@ func TestRelease(t *testing.T) {
 			if !AreErrorsEqual(releaseErr, tc.expectErr) {
 				t.Fatalf("Got error %v, expected error %v", releaseErr, tc.expectErr)
 			}
+			if tc.expectedRes != nil && tc.expectedRes.Status.UserData == nil {
```
Comment (ixdy) [inline boskos/ranch/ranch_test.go]: couldn't we fix this by making `crds.NewResource` always set UserData to an empty struct?

**Comment by ixdy** [comment]
I clicked the wrong button, but what I also meant to add is that this is really cool! and this change also helps illustrate that we really need to refactor a lot of the testing code into a `boskos/testing` package or something like that.

File: boskos/ranch/storage.go [hunk 3]
```diff
@@ -85,7 +86,7 @@ func NewStorage(ctx context.Context, client ctrlruntimeclient.Client, namespace,
 		}
 
 		for _, res := range data.Resources {
-			if err := s.AddResource(res); err != nil {
+			if err := s.AddResource(crds.FromResource(res)); err != nil {
```
[the remark this answers is not in the document]
Comment (alvaroaleman) [reply boskos/ranch/storage.go]: Yeah, good point. It also doesn't really make sense conceptionally, if I load my state from disk I also want to save it there, but that is currently not possible.

Are you good with removing this in a follow-up to keep the size of this manageable?

File: boskos/ranch/storage.go [hunk 4]
```diff
@@ -114,10 +114,11 @@ func (s *Storage) DeleteResource(name string) error {
 
 // UpdateResource updates a resource if it exists, errors otherwise
 func (s *Storage) UpdateResource(resource *crds.ResourceObject) (*crds.ResourceObject, error) {
+	resource.Namespace = s.namespace
```
[the remark this answers is not in the document]
Comment (alvaroaleman) [reply boskos/ranch/storage.go]: Well we need something that enforces the namespace. Another approach could be to add a wrapper around the client itself, but I found it more logical in the storage, because it must be configured with a namespace anyways, as it has some `Get(name string)` funcs

File: boskos/ranch/storage.go [hunk 5]
```diff
@@ -114,60 +113,45 @@ func (s *Storage) DeleteResource(name string) error {
...
-	if err := s.client.Get(s.ctx, name, o); err != nil {
-		return common.Resource{}, fmt.Errorf("failed to get resource %s before patching it: %v", resource.GetName(), err)
-	}
+func (s *Storage) UpdateResource(resource *crds.ResourceObject) (*crds.ResourceObject, error) {
+	resource.Status.LastUpdate = s.now()
 
-	o.FromItem(resource)
-	if err := s.client.Update(s.ctx, o); err != nil {
-		return common.Resource{}, fmt.Errorf("failed to update resources %s after patching it: %v", resource.GetName(), err)
+	if err := s.client.Update(s.ctx, resource); err != nil {
```
[the remark this answers is not in the document]
Comment (alvaroaleman) [reply boskos/ranch/storage.go]: We are more likely to fail, but thanks to the global lock still not very likely. I will definitely add retrying in both the server and the client, but I would like to do that after this PR merged, to I do not have to rebase so much. Would that be okay?

File: boskos/ranch/ranch_test.go (hunk 2 above)
Comment (alvaroaleman) [reply boskos/ranch/ranch_test.go]: Yeah, but then we need to set it on other tests so they keep passing :) I did it anyways, mostly because the code seems to assume that its not nil and calls methods on it that will just panic if its nil. I did it in a separate commit, tell me what you think.

File: boskos/crds/client.go [hunk 6]
```diff
@@ -153,6 +153,9 @@ type Type struct {
 
 // Object extends the runtime.Object interface. CRD are just a representation of the actual boskos object
 // which should implements the common.Item interface.
+// TODO alvaroaleman: Remove this interface and its implementations, the only thing it adds on top of
```
[the remark this answers is not in the document]
Comment (alvaroaleman) [reply boskos/crds/client.go]: Yeah exactly, the PR has to be huge already and I didn't want to make that even worse. I've created an issue and updated the code comment to link to that instead.

**Comment by alvaroaleman** [comment]
> I clicked the wrong button, but what I also meant to add is that this is really cool! and this change also helps illustrate that we really need to refactor a lot of the testing code into a boskos/testing package or something like that.

Yeah, a lot of the comparing could be commoditized and I am also reasonable sure there is a decent amount of dead code

**Comment by alvaroaleman** [comment]
This was sitting for almost a week, I think it should be fine now.
/hold cancel

File: boskos/ranch/storage.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ixdy) [reply boskos/ranch/storage.go]: created a follow-up PR to clean this up: #16694
